### PR TITLE
Added multilabel stratification to AbsTaskMultilabelClassification

### PR DESCRIPTION
**Pull request by dokato**
This is continuation of the discussion from https://github.com/embeddings-benchmark/mteb/discussions/698. 

cc @x-tabdeveloping

### CODE REVIEW

File: mteb/abstasks/AbsTaskMultilabelClassification.py [hunk 1]
```diff
@@ -36,6 +38,20 @@ def evaluate_classifier(
...
+    label_set = sorted(label_set)
+
+    multihot_vectors = []
+    for label_list in labels:
+        multihot_vectors.append([1 if x in label_list else 0 for x in label_set])
+
+    return np.array(multihot_vectors)
+
+
 class AbsTaskMultilabelClassification(AbsTask):
```
Comment (x-tabdeveloping) [inline mteb/abstasks/AbsTaskMultilabelClassification.py:55]: The problem is that this behaviour is not only needed in tasks that always have multiple labels. I implemented this behaviour in AbsTask originally as a backoff for when single label stratification fails, I think we should probably go for that approach.

File: pyproject.toml [hunk 2]
```diff
@@ -40,6 +40,7 @@ dependencies = [
     "typing_extensions",
     "eval_type_backport",
     "polars>=0.20.22",
+    "skmultilearn",
```
Comment (x-tabdeveloping) [inline pyproject.toml:43]: Have you had a look at how this iterative subsampling function works in skmultilearn? If we can implement it ourselves, I would probably prefer that over introducing a new dependency. If it would be too complex to implement in MTEB we can of course roll with sklmultilearn.

**Review by x-tabdeveloping — COMMENTED** [review]
See my comments. Let's continue discussion on this, I think we are going in the right direction but this still needs some thought and work.

File: mteb/abstasks/AbsTaskMultilabelClassification.py [hunk 3]
```diff
@@ -36,6 +38,20 @@ def evaluate_classifier(
     return scores
 
 
+def _multihot_encoder(labels):
```
[the remark this answers is not in the document]
Comment (dokato) [reply mteb/abstasks/AbsTaskMultilabelClassification.py]: yup, removed

File: mteb/abstasks/AbsTaskMultilabelClassification.py (hunk 1 above)
Comment (dokato) [reply mteb/abstasks/AbsTaskMultilabelClassification.py]: I don't quite get it why we should even try single label stratification on a multilabel problem?

File: pyproject.toml (hunk 2 above)
Comment (dokato) [reply pyproject.toml:43]: Yea, I had a quick look. It's not terribly complicated but there's a few non-trivial things going on, see:
- https://github.com/scikit-multilearn/scikit-multilearn/blob/master/skmultilearn/model_selection/iterative_stratification.py#L175
- ["A Network Perspective on Stratification of Multi-Label Data", Szymański & Kajdanowicz, 2017](http://proceedings.mlr.press/v74/szyma%C5%84ski17a/szyma%C5%84ski17a.pdf)

**Comment by KennethEnevoldsen** [comment]
hi @dokato and @x-tabdeveloping: I would probably just do the split when creating the dataset on HF. This avoids the extra dependency.

**Comment by x-tabdeveloping** [comment]
Yeah but then we need to reupload a lot of datasets, and I'm wondering if it's too much of a hustle. I am really in favour of not introducing new dependencies if possible.
I'm wondering if it would make sense to just copy the file over from sklmultilearn, simplify it a bit, and call it a day.
It seems pretty self-contained to me at least. What do you think @dokato @KennethEnevoldsen ?

**Comment by KennethEnevoldsen** [comment]
@x-tabdeveloping I believe that is the best solution. We just need to add a reference

**Comment by dokato** [comment]
Thanks for advise @x-tabdeveloping @KennethEnevoldsen. That's what I did, I moved their function (with acknowledgments) and slightly modified so it returns indices `_iterative_train_test_split`.

File: mteb/abstasks/AbsTaskMultilabelClassification.py (hunk 1 above)
Comment (dokato) [reply mteb/abstasks/AbsTaskMultilabelClassification.py]: Cool, thanks for elaborating. I see your points. If useful in other tasks, it should definately be in AbsTask. Though, the place I don't like are those conditions:

```
            try:
                dataset_dict = dataset_dict.class_encode_column(label)
            except ValueError as e:
                if isinstance(dataset_dict[splits[0]][label][0], Sequence):
                    ...
```
IMHO we should just recognise by task type and apply stratified split where it really makes sense, instead of relying on try/except.

Comment (x-tabdeveloping) [reply mteb/abstasks/AbsTaskMultilabelClassification.py]: We can't do it, cause ClusteringFast is both hierarchical and flat unfortunately.

**Comment by x-tabdeveloping** [comment]
Sorry for the delay @dokato I'm just mid exam-season, I'll have a look at it right now :D

**Review by x-tabdeveloping — COMMENTED** [review]
Beautiful! Looks absolutely good to me. Can I ask you to implement it for one or more tasks and see if it works? Thanks for your awesome work and patience

**Comment by dokato** [comment]
Thanks for review @x-tabdeveloping! I rerun Brazilian dataset as all other had small enough "test" splits that didn't require that much of stratification. Hence, I added additionally `KorHateSpeechMLClassification` (https://huggingface.co/datasets/jeanlee/kmhas_korean_hate_speech) also to improve representation of MultiLabel Classification task.

Here's dataset card for `KorHateSpeechMLClassification`.

- [x] I have tested that the dataset runs with the `mteb` package.
- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [x] I have added points for my submission to the [points folder](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) using the PR number as the filename (e.g. `438.jsonl`).

**Comment by dokato** [comment]
I've changed "Stratified subsampling of test set to 2000 examples." bit to more commonly used in classification 2048 `AbsTaskMultilabelClassification.py`. But now, given that we have proper stratificaiton I wonder if we need that bit at all, or should we just assume that the sampling is done in the `dataset_transform` method? Similarly as is in other tasks. Maybe just a warning if the split if bigger than 2048 would suffice?

**Comment by dokato** [comment]
@x-tabdeveloping struggling to understand what might be going on here. I tried to rebase but it had too many merge conflicts, so I forced push it. But honestly, have no idea why those tests are failing, esp. that they related to some other datasets...

**Comment by x-tabdeveloping** [comment]
Well, since this PR is making changes to quite fundamental things in the library, I would prefer if everything was passing before we went on, and all conflicts were resolved. I think if nothing fixes it probably the best thing to do is just reapply your changes to the current main in a new PR or a new branch, it seems like the most painless option to me. I don't have the slightest cue either what might have gone wrong here unfortunately.

**Comment by KennethEnevoldsen** [comment]
@dokato pulling from main should resolve a lot of the dataset issues.

**Comment by dokato** [comment]
Good shoutout guys, I rebased again. I guess it's ready?

**Comment by x-tabdeveloping** [comment]
I think it looks alright! Thanks for the work and patience @dokato :D

**Comment by dokato** [comment]
No worries, pleasure! hope you're exams went fine ;)
